### ISSUE TITLE
Angular 9 - Not unmounting

### ISSUE THREAD

**Issue by kyro38** [single-spa/single-spa-angular#130]
Since upgrade to Angular 9, I've noticed that the angular app in not unmounting correctly.

When the route is changed and no app is replacing the angular app, the angular doesn't disappear. But it gets more weird. The `(click)` methods are not called. 

So some unmounting is happening but the app stays in the DOM. 

It seems, I'm [not the only one who](https://github.com/single-spa/single-spa-angular/issues/128#issuecomment-585124332) experiences this problem with Angular 9.

**Comment by joeldenning** [comment]
👍 thanks for reporting this. I hope to tackle all Angular 9 issues in the next few weeks.

**Comment by riya332** [comment]
I was stuck over 2 days in the same situation. I tried several things but none of them worked. I thought it was a problem with angular routing.  Finally, realized only after downgrading to angular @ 8 , That the problem exists only in A9 apps. I have downgraded to A8 and changed the version of     **"@angular-devkit/build-angular":  "^0.803.24",** It works for me.

**Comment by joeldenning** [comment]
I'm not sure when I'll get to Angular 9 support - I am not very strong in Angular and usually ZoneJS + Angular Router kick my ass when it comes to making Angular work with single-spa. I'd very gladly accept pull requests that do this.

**Comment by richardbladh** [comment]
A very naive suggestion is to add the following line at the end of the function unmount in single-spa-angular.ts
It seems to remove the DOM at least but I'm not sure what other things might still be there...
getContainerEl(opts.domElementGetter).innerHTML = '';

**Comment by joeldenning** [comment]
@richardbladh thanks for jumping in on this issue - you are correct that would sort of work but the problem is that the `onDestroy` lifecycles would not be called in that situation.

Currently we unmount angular by doing the following:

https://github.com/single-spa/single-spa-angular/blob/5f7b1604d0fbdb42d0dc950538907e5dbfef43dd/src/browser-lib/single-spa-angular.ts#L110

Where bootstrappedModule is the result of calling the bootstrapFunction. I haven't looked into why it's failing with Angular 9, but hopefully we wouldn't have to deviate too far from what we're already doing in the code I linked to above.

**Comment by robinvdbroek** [comment]
There seems to be a different execution of the bootstrap function, looking from debugger the bootstrappedModule has type `NgModuleRef$1` in Angular 9, vs. `NgModuleRef_` in Angular 8, which is probably due to Ivy's new rendering engine. 

Also noteworthy is an error that occurs on line 130 of single-spa-angular.js where the `animationEngine` is resolved from an injector (`opts.bootstrappedModule.injector.get(opts.AnimationEngine)`. The errors says the injector has already been destroyed (which could be done by line 110 ofc.).

**Comment by joeldenning** [comment]
> Also noteworthy is an error that occurs on line 130 of single-spa-angular.js where the animationEngine is resolved from an injector (opts.bootstrappedModule.injector.get(opts.AnimationEngine). The errors says the injector has already been destroyed (which could be done by line 110 ofc.).

That is good news - single-spa-angular was working around a bug in the animation engine, but from what you said it appears that that might be fixed as of Angular 9. If so, the single-spa-angular browser lib should continue doing the workaround for old versions of angular, but no longer do it for Angular 9. Related code (click to see full comment):

https://github.com/single-spa/single-spa-angular/blob/5f7b1604d0fbdb42d0dc950538907e5dbfef43dd/src/browser-lib/single-spa-angular.ts#L113

**Comment by richardbladh** [comment]
Ok, so I ended up with a temporary solution where I implement OnDestroy in my app component.
Then in the ngOnDestroy method I reuse the domElementGetter-function that I provided in the app registration in single spa to find the root element. Then I just set innerHTML of that element to an empty string.

It works for now as a "solution" for me since it will only affects the Angular application that contains that code (my new Angular 9 app) without affecting my old Angular apps.

**Comment by arturovt** [comment]
It's a bug of the Ivy engine. Reported to Angular team:

https://github.com/angular/angular/issues/36449

**Comment by orosro** [comment]
> Ok, so I ended up with a temporary solution where I implement OnDestroy in my app component.
> Then in the ngOnDestroy method I reuse the domElementGetter-function that I provided in the app registration in single spa to find the root element. Then I just set innerHTML of that element to an empty string.
> 
> It works for now as a "solution" for me since it will only affects the Angular application that contains that code (my new Angular 9 app) without affecting my old Angular apps.

@richardbladh  Could you provide an example on how you did it? Thank you

**Comment by richardbladh** [comment]
In app.component.ts in your Angular 9 app.
`export class AppComponent implements OnDestroy {
    ngOnDestroy() {
        // Fix for Angular 9 + SingleSpa. Remove when the issue is fixed: https://github.com/single-spa/single-spa-angular/issues/130
        // Solution (seems to work for now): Cleanup the DOM manually on destroy.
        const element = getTheElementFromSomeGlobalRegistryThatYouveSetupWhenYouRegistreredYourApplication();
        element.innerHTML = "";
    }
}`

**Comment by orosro** [comment]
Thank you,

I'm working on a POC and I'm going to use this solution for the moment. I saw that there are also some errors with the AnimationEngine, removing it from the singleSpaAngular bootstrap method "fixed" the unmounting process.

**Comment by arturovt** [comment]
@orosro 

if you've encountered any issues you might want to open an issue and _the most important_ to provide a _minimal reproducible example_. I will try to help you to resolve those issues.

**Comment by JasonEdLewis** [comment]
> In app.component.ts in your Angular 9 app.
> 
> `export class AppComponent implements OnDestroy {
> 
> ```
> ngOnDestroy() {
>     // Fix for Angular 9 + SingleSpa. Remove when the issue is fixed: https://github.com/single-spa/single-spa-angular/issues/130
>     // Solution (seems to work for now): Cleanup the DOM manually on destroy.
>     const element = getTheElementFromSomeGlobalRegistryThatYouveSetupWhenYouRegistreredYourApplication();
>     element.innerHTML = "";
> }
> ```
> 
> }`

@richardbladh Does this assume that the element was placed on the DOM using the DomElementGetter ?

**Comment by filoxo** [comment]
Looks like the fix for this happened in https://github.com/angular/angular/pull/37600 and landed in [Angular 10.0.1](https://github.com/angular/angular/blob/master/CHANGELOG.md#1001-2020-06-26).

**Comment by sneeldeshmukh** [comment]
Tried with Angular [10.0.10 ](https://github.com/angular/angular/blob/master/CHANGELOG.md#1001-2020-06-26) still the issue persists

**Comment by sneeldeshmukh** [comment]
And If it's fixed in  Angular 10.0.1. so is it now to be also fix from single-spa library as well (in any next version) , and If yes so is there any road-map will be helpful if someone can share insides on this.

**Comment by joeldenning** [comment]
My understanding is that this issue is not with single-spa itself, but with Angular. Single-spa calls `lifecycles.unmount()`, which is implemented here in single-spa-angular:

https://github.com/single-spa/single-spa-angular/blob/91b1513fa1db69851acfd8d0b7904cca18a0bbe7/src/src/single-spa-angular.ts#L157

The main thing that occurs in that method is that we call `.destroy()` on the mounted Angular module:

https://github.com/single-spa/single-spa-angular/blob/91b1513fa1db69851acfd8d0b7904cca18a0bbe7/src/src/single-spa-angular.ts#L197

Additionally, since Ivy had a bug we have a workaround that removes it from the dom:

https://github.com/single-spa/single-spa-angular/blob/91b1513fa1db69851acfd8d0b7904cca18a0bbe7/src/src/single-spa-angular.ts#L202

Ideally we wouldn't have to do any workarounds / hacks, but we can add any things needed to that `unmount` function that help things unmount.

**Comment by sneeldeshmukh** [comment]
Already added in  **unmount** function
`singleSpa.unloadApplication(appName);`

Also the above function gets executed but the app doesn't fully unload and stay inside the Mounted Apps array 
ie. by calling bellow function after unload I can see the app is still there 
`singleSpa.getMountedApps()`

**Comment by joeldenning** [comment]
You should not call `singleSpa.unloadApplication()` unless you want to **redownload the source code for the application**. This is often used for hot reloading and is not related to unmounting. I think you might be using it by mistake, as it has nothing to do with what is described in this github issue.

> Also the above function gets executed but the app doesn't fully unload and stay inside the Mounted Apps array

You have to wait for `unloadApplication` to finish before you call `singleSpa.getMountedApps()`. For example:

```js
singleSpa.unloadApplicaton(appName).then(() => {
  console.log(singleSpa.getMountedApps())
})
```

**Comment by arturovt** [comment]
Hey guys. I want to clarify things. I've checked that the application is removed successfully w/o using any hacks like `element.remove()`, etc.

P.S. I've checked with `angular@10.2.0`. The fix was published with `10.0.1` version: https://github.com/angular/angular/blob/master/CHANGELOG.md#1001-2020-06-26 (core: cleanup DOM elements when root view is removed).

That means that this function is not needed anymore: https://github.com/single-spa/single-spa-angular/blob/master/src/internals/src/dom.ts#L3-L15

Also it doesn't work correctly when using SystemJS, since `ivyEnabled()` always returns false (because it throws `require` is not defined at runtime).
